### PR TITLE
ODF actor implemented with Ray Tracing

### PR DESCRIPTION
**Pull request by tvcastillod**
This PR aims to create a new odf actor defined with raymarching based on the paper _[Ray Tracing Spherical Harmonics Glyphs](https://www.willusher.io/publications/rt-sph-glyphs)_ . Work is still in progress.

A visual comparison with the current implementation of fury using polygons
image

An [example](https://github.com/fury-gl/fury/blob/d9c332340bb579e28f9c59706ec7442c0c7b4408/docs/experimental/SH-ODF%20experimental/odf_example.py) is provided showing that it supports displaying multiple glyphs, and with different SH degree.
image

### CODE REVIEW

File: fury/actor.py [hunk 1]
```diff
@@ -3954,3 +3955,82 @@ def uncertainty_cone(
...
+        coeffs = np.array([coeffs])
+    if coeffs.shape[0] != centers.shape[0]:
+        raise ValueError('number of odf glyphs defined does not match with '
+                         'number of centers')
+
+    coeffs_given = coeffs.shape[-1]
+    if degree is None:
+        degree = int((np.sqrt(8 * coeffs_given + 1) - 3)/2)
+    elif degree % 2 != 0:
+        raise ValueError('degree must be even')
```
Comment (guaje) [inline fury/actor.py:4014]: `warnings.warn('Invalid degree value. Degree must be a positive even number lower or equal to 12. Ignoring passed value and using maximum degree supported by the number of SH coefficients.')`

File: fury/actor.py [hunk 2]
```diff
@@ -3954,3 +3955,82 @@ def uncertainty_cone(
...
+        coeffs = np.array(coeffs)
+    if coeffs.ndim == 1:
+        coeffs = np.array([coeffs])
+    if coeffs.shape[0] != centers.shape[0]:
+        raise ValueError('number of odf glyphs defined does not match with '
+                         'number of centers')
+
+    coeffs_given = coeffs.shape[-1]
+    if degree is None:
+        degree = int((np.sqrt(8 * coeffs_given + 1) - 3)/2)
```
Comment (guaje) [inline fury/actor.py:4012]: For comparison purposes this needs to be calculated anyway, so precalculate it before the `if` validations. The logic should be similar to this:

```python
max_degree = int((np.sqrt(8 * coeffs_given + 1) - 3) / 2)
if degree is None:
    degree = max_degree
else:
    if degree % 2 != 0:
        warnings.warn('Invalid degree value. Degree must be a positive even number ...')
        degree = max_degree
    else:
        coeffs_needed = ((degree + 1) * (degree + 2)) / 2
        if coeffs_given < coeffs_needed:
            warnings.warn('Not enough number of SH coefficients for degree {0}...')
            degree = max_degree
```

**Review by guaje — CHANGES_REQUESTED** [review]
Hi @tvcastillod, this new actor looks amazing!

I have added some comments to your PR. I'll keep adding comments to it.

File: fury/actors/odf.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        );
+        """
+
+    # Identify the first intersection
+    first_intersection = \
+        """
+        float first_ray_param = NO_INTERSECTION;
+        _unroll_
+        for (int i = 0; i != MAX_DEGREE; ++i) {
+        //for (int i = 0; i != maxPolyDegreeVSOutput; ++i) {
```
Comment (guaje) [inline fury/actors/odf.py:358]: Use the line with the uniform polynomial degree.

File: fury/actors/odf.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+
+    # Evaluate shading for a directional light
+    directional_light = \
+        """
+        vec3 color = vec3(1.);
+        if (first_ray_param != NO_INTERSECTION) {
+            vec3 intersection = ro - centerMCVSOutput + first_ray_param * rd;
+            vec3 normal = getShGlyphNormal(sh_coeffs, intersection, int(shDegree), int(numCoeffsVSOutput));
+            vec3 colorDir = srgbToLinearRgb(abs(normalize(intersection)));
+            float attenuation = ld.z;//dot(ld, normal);
```
Comment (guaje) [inline fury/actors/odf.py:374]: Use `dot(ld, normal)`

**Review by guaje — CHANGES_REQUESTED** [review]
Hello @tvcastillod, here are some minor comments on some of these files.

File: fury/shaders/utils/find_roots.frag [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    derivative[0] = poly[maxPolyDegree - 2];
+    derivative[1] = float(maxPolyDegree - 1) * poly[maxPolyDegree - 1];
+    derivative[2] = (0.5 * float((maxPolyDegree - 1) * maxPolyDegree)) * poly[maxPolyDegree - 0];
+    _unroll_
+    for (int i = 3; i != maxPolyDegree + 1; ++i)
+        derivative[i] = 0.0;
+    // Compute its two roots using the quadratic formula
+    float discriminant = derivative[1] * derivative[1] - 4.0 * derivative[0] * derivative[2];
+    if (discriminant >= 0.0) {
+        float sqrt_discriminant = sqrt(discriminant);
```
Comment (guaje) [inline fury/shaders/utils/find_roots.frag:18]: Refactor using camel case convention (`sqrtDiscriminant`)

File: fury/shaders/utils/find_roots.frag [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        float scaled_root = derivative[1] + ((derivative[1] > 0.0) ? sqrt_discriminant : (-sqrt_discriminant));
```
Comment (guaje) [inline fury/shaders/utils/find_roots.frag:19]: Refactor using camel case convention (`scaledRoot`)

File: fury/shaders/utils/find_roots.frag [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        float root_0 = clamp(-2.0 * derivative[0] / scaled_root, begin, end);
```
Comment (guaje) [inline fury/shaders/utils/find_roots.frag:20]: Refactor using camel case convention (`root0`)

File: fury/shaders/utils/find_roots.frag [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        float root_1 = clamp(-0.5 * scaled_root / derivative[2], begin, end);
```
Comment (guaje) [inline fury/shaders/utils/find_roots.frag:21]: Refactor using camel case convention (`root1`)

File: fury/shaders/utils/find_roots.frag [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    // constructs the derivatives in a contrived way. Changing that would
+    // reduce the number of arithmetic instructions roughly by a factor of two.
+    // However, it would also cause register spilling, which has a far more
+    // negative impact on the overall run time. Profiling indicates that the
+    // current implementation has no spilling whatsoever.
+    _loop_
+    for (int degree = 3; degree != maxPolyDegree + 1; ++degree) {
+        // Take the integral of the previous derivative (scaled such that the
+        // constant coefficient can still be copied directly from poly)
+        float prev_derivative_order = float(maxPolyDegree + 1 - degree);
```
Comment (guaje) [inline fury/shaders/utils/find_roots.frag:48]: Refactor using camel case convention (`prevDerivativeOrder`)

File: fury/shaders/utils/find_roots.frag [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        _unroll_
+        for (int i = maxPolyDegree; i != 0; --i)
+            derivative[i] = derivative[i - 1] * (prev_derivative_order * (1.0 / float(i)));
+        // Copy the constant coefficient without causing spilling. This part
+        // would be harder if the derivative were not scaled the way it is.
+        _unroll_
+        for (int i = 0; i != maxPolyDegree - 2; ++i)
+            derivative[0] = (degree == maxPolyDegree - i) ? poly[i] : derivative[0];
+        // Determine the value of this derivative at begin
+        float begin_value = derivative[maxPolyDegree];
```
Comment (guaje) [inline fury/shaders/utils/find_roots.frag:58]: Refactor using camel case convention (`beginValue`)

File: fury/shaders/utils/find_roots.frag [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        _unroll_
+        for (int i = maxPolyDegree - 1; i != -1; --i)
+            begin_value = begin_value * begin + derivative[i];
+        // Iterate over the intervals where roots may be found
+        _unroll_
+        for (int i = 0; i != maxPolyDegree; ++i) {
+            if (i < maxPolyDegree - degree)
+                continue;
+            float current_begin = outRoots[i];
```
Comment (guaje) [inline fury/shaders/utils/find_roots.frag:67]: Refactor using camel case convention (`currentBegin`)

File: fury/shaders/utils/find_roots.frag [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+            float current_end = outRoots[i + 1];
```
Comment (guaje) [inline fury/shaders/utils/find_roots.frag:68]: Refactor using camel case convention (`currentEnd`)

File: fury/shaders/rt_odfs/tournier/eval_sh_8.frag [hunk 13]
```diff
@@ -0,0 +1,122 @@
+void eval_sh_8(out float outSH[45], vec3 point)
```
Comment (guaje) [inline fury/shaders/rt_odfs/tournier/eval_sh_8.frag:1]: Refactor using camel case convention (`evalSH8`)

File: fury/shaders/rt_odfs/tournier/eval_sh_6.frag [hunk 14]
```diff
@@ -0,0 +1,79 @@
+void eval_sh_6(out float outSH[28], vec3 point)
```
Comment (guaje) [inline fury/shaders/rt_odfs/tournier/eval_sh_6.frag:1]: Refactor using camel case convention (`evalSH6`)

File: fury/shaders/rt_odfs/tournier/eval_sh_4.frag [hunk 15]
```diff
@@ -0,0 +1,46 @@
+void eval_sh_4(out float outSH[15], vec3 point)
```
Comment (guaje) [inline fury/shaders/rt_odfs/tournier/eval_sh_4.frag:1]: Refactor using camel case convention (`evalSH4`)

File: fury/shaders/rt_odfs/tournier/eval_sh_2.frag [hunk 16]
```diff
@@ -0,0 +1,23 @@
+void eval_sh_2(out float outSH[6], vec3 point)
```
Comment (guaje) [inline fury/shaders/rt_odfs/tournier/eval_sh_2.frag:1]: Refactor using camel case convention (`evalSH2`)

File: fury/shaders/rt_odfs/tournier/eval_sh_12.frag [hunk 17]
```diff
@@ -0,0 +1,238 @@
+void eval_sh_12(out float outSH[91], vec3 point)
```
Comment (guaje) [inline fury/shaders/rt_odfs/tournier/eval_sh_12.frag:1]: Refactor using camel case convention (`evalSH12`)

File: fury/shaders/rt_odfs/tournier/eval_sh_10.frag [hunk 18]
```diff
@@ -0,0 +1,175 @@
+void eval_sh_10(out float outSH[66], vec3 point)
```
Comment (guaje) [inline fury/shaders/rt_odfs/tournier/eval_sh_10.frag:1]: Refactor using camel case convention (`evalSH10`)

File: fury/shaders/rt_odfs/tonemap.frag [hunk 19]
```diff
@@ -0,0 +1,5 @@
+vec3 tonemap(vec3 linear)
+{
+    float max_channel = max(max(1.0, linear.r), max(linear.g, linear.b));
```
Comment (guaje) [inline fury/shaders/rt_odfs/tonemap.frag:3]: Refactor using camel case convention (`maxChannel`)

File: fury/actors/odf.py [hunk 20]
```diff
@@ -0,0 +1,395 @@
...
+        os.path.join("lighting", "linear_rgb_to_srgb.frag")
+    )
+
+    # Inverse of linear_rgb_to_srgb()
+    srgb_to_linear_rgb = import_fury_shader(
+        os.path.join("lighting", "srgb_to_linear_rgb.frag")
+    )
+
+    # Logarithmic tonemapping operator. Input and output are linear RGB.
+    tonemap = import_fury_shader(os.path.join("rt_odfs", "tonemap.frag"))
```
Comment (guaje) [inline fury/actors/odf.py:293]: Move to `lighting`

File: fury/shaders/rt_odfs/eval_sh.frag [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+        else if (shDegree == 12)
+        {
+            float tmpOutSH[91];
+            #if SH_DEGREE == 12
+                evalSH12(tmpOutSH, point);
+            #endif
+            for (int i = 0; i != numCoeffs; ++i)
+                outSH[i] = tmpOutSH[i];
+        }
+    }
```
Comment (guaje) [inline fury/shaders/rt_odfs/eval_sh.frag:58]: Fix indentation and add an empty line at the end of the file

**Review by guaje — CHANGES_REQUESTED** [review]
Hello @tvcastillod!

Overall great work putting up together such a needed actor.

In addition to the changes outlined below, I suggest moving the content of `rt_odfs` to `ray_tracing/odf`.

File: fury/actor.py [hunk 22]
```diff
@@ -4019,3 +4020,73 @@ def uncertainty_cone(
     angles = main_dir_uncertainty(evals, evecs, signal, sigma, b_matrix)
 
     return double_cone(centers, evecs, angles, colors, scales, opacity)
+
+
+@warn_on_args_to_kwargs()
```
Comment (skoudoro) [inline fury/actor.py:4025]: not needed since this is a new function

File: fury/actor.py [hunk 23]
```diff
@@ -4019,3 +4020,73 @@ def uncertainty_cone(
...
+    centers : ndarray(N, 3)
+        ODFs positions.
+    coeffs : (N, M) or (N, 6) or (N, 15) or (N, 28) or (N, 45) or (N, 66) or
+        (N, 91) ndarray.
+        Corresponding SH coefficients for the ODFs.
+    sh_basis: str, optional
+        Type of basis (descoteaux, tournier)
+        'descoteaux' for the default ``descoteaux07`` DYPY basis.
+        'tournier' for the default ``tournier07` DYPY basis.
+    scales : float or ndarray (N, )
```
Comment (skoudoro) [inline fury/actor.py:4049]: `, optional` to add

File: fury/actor.py [hunk 24]
```diff
@@ -4019,3 +4020,73 @@ def uncertainty_cone(
...
+        ODFs size.
+    opacity : float
```
Comment (skoudoro) [inline fury/actor.py:4051]: `, optional` to add

File: fury/utils.py [hunk 25]
```diff
@@ -1622,3 +1622,36 @@ def set_actor_origin(actor, *, center=None):
...
+
+
+def minmax_norm(data, axis=1):
+    """Returns the min-max normalization of data along an axis.
+
+    Parameters
+    ----------
+    data: ndarray
+        2D array
+    axis: int
```
Comment (skoudoro) [inline fury/utils.py:1634]: `, optional` to add

**Review by skoudoro — COMMENTED** [review]
exciting, thank you for this work @tvcastillod !

First, can you fix the issues from code format CI's and build doc CI's

then, see my first initial comments below.

Question: Can you explain a bit how did you get all this `eval_sh_*`?

**Comment by tvcastillod** [comment]
> Question: Can you explain a bit how did you get all this `eval_sh_*`?

Hi @skoudoro,

The `eval_sh_*` are taken directly from the source code given in [the paper](https://www.willusher.io/publications/rt-sph-glyphs/) I took as reference, you can check their [shadertoy demo](https://www.shadertoy.com/view/dlGSDV) for a quick look. These values correspond to the SH basis functions listed in _Appendix A_ from the paper, which comes from the definition of Spherical Harmonics, the same from the [DIPY documentation](https://grgweb.luddy.indiana.edu/documentation/1.7.0/theory/sh_basis/). What I added was the Tournier basis which is very similar to Descoteaux with some sign changes.

**Comment by skoudoro** [comment]
Hi @tvcastillod, 

Overall, it is almost ready to be merged. Can you fix Codespell CI's and Docs CI's which are failing...

In the docs CI, you have this 2 warnings:
```python
/home/runner/work/fury/fury/fury/actor.py:docstring of fury.actor.odf:16: WARNING: Inline literal start-string without end-string.
/home/runner/work/fury/fury/fury/actors/odf.py:docstring of fury.actors.odf.sh_odf:14: WARNING: Inline literal start-string without end-string.
```

Concerning the codespell, you need to add `shs` as an ignore word to avoid the issue. 


Thank you for the future update

**Comment by maharshi-gor** [comment]
Hello @tvcastillod @guaje  @skoudoro 

Thanks for the work!

I tested this PR with a large dataset and when I tried to load the whole slice of that dataset. It worked perfectly fine. But if I zoom in it becomes distorted.
unnamed

**NOTE**

1. The current version of fury ODF Slicer Actor was smooth and better. I am only adding one slice of data to the proposed odf actor as well.
2. The current version of fury ODF Slicer Actor does not distort the image as well.

**My Observation and Take away**

- I think it is something related to zoom and the shapes are not getting recalculated once zoomed in more.
- We are doing something not right in the calculation implementation as it should be faster than the approach we have in fury.

Let me know if you need any information.

**Comment by guaje** [comment]
> Hello @tvcastillod @guaje @skoudoro
> 
> Thanks for the work!
> 
> I tested this PR with a large dataset and when I tried to load the whole slice of that dataset. It worked perfectly fine. But if I zoom in it becomes distorted. unnamed
> 
> **NOTE**
> 
> 1. The current version of fury ODF Slicer Actor was smooth and better. I am only adding one slice of data to the proposed odf actor as well.
> 2. The current version of fury ODF Slicer Actor does not distort the image as well.
> 
> **My Observation and Take away**
> 
> * I think it is something related to zoom and the shapes are not getting recalculated once zoomed in more.
> * We are doing something not right in the calculation implementation as it should be faster than the approach we have in fury.
> 
> Let me know if you need any information.

Thanks for this thorough review @maharshi-gor. Could you share the dataset and detailed steps to generate the error?

**Comment by maharshi-gor** [comment]
Hello @guaje @tvcastillod,

It is a dataset called setsompop. Please find the Dataset [here](https://indiana-my.sharepoint.com/:f:/g/personal/magor_iu_edu/Eiwiu6jfVYlLjYQVIUtl2hoBVA1Wuf4lD-cq1qMDTT3cBQ?e=Ogjqm2)

Please fit the entire slice using CsaOdfModel and then zoom it.

Also, I would love to see an example from your-side to use the new odf actor with a odf model as the inputs are different from the current odf slicer actor. This might be an input issue!

Thanks.

**Comment by maharshi-gor** [comment]
Hey @skoudoro 

Let's merge this PR as for all the normal dataset it works fine. But for the large dataset @guaje and @tvcastillod are working on it.

I have created another issue to track that [here](https://github.com/fury-gl/fury/issues/943).

**Comment by skoudoro** [comment]
Thank @maharshi-gor for the feedback, thanks @tvcastillod for this work and thanks @guaje for the review. merging

**Pull request merged**